### PR TITLE
Reset the current progress tracking state during double evaluation

### PR DESCRIPTION
**Pull request by carmocca**
## What does this PR do?

With fault-tolerance disabled, the following runs twice through the test set, it's equivalent to `max_test_epochs=2`

```python
trainer = Trainer(...)
trainer.test(...)
trainer.test(...)
```

However, currently in master, if fault-tolerance is enabled, the second test exits early. This is because the `done` condition for the test loop has been already fulfilled:

```py
self.dataloader_progress.current.completed >= self.num_dataloaders
```

To avoid this, we need to check whether we are NOT fitting, and whether the previous state is done, then in that case reset the state.

This is different from how fit works where this is the expected behavior and the user is required to set a different `max_epochs` value. The equivalent would be that the user has to change `self.num_dataloaders` which does not make sense.

Predict also has this bug, but there we don't need the fit check.

### Test plan

This is covered by existing tests, for example: `test_checkpoint_path_input`

### Does your PR introduce any breaking changes? If yes, please list them.

None

## Before submitting

- [n/a] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [n/a] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

cc @carmocca @justusschock @awaelchli @ninginthecloud @borda

### CODE REVIEW

File: pytorch_lightning/loops/dataloader/evaluation_loop.py [hunk 1]
```diff
@@ -80,6 +81,10 @@ def reset(self) -> None:
             self._max_batches = [self._max_batches] * len(self.dataloaders)
 
         super().reset()
+        # when restarting, if we are running `validate` or `test` twice, since there's no concept of `max_epochs` we
+        # need to reset the current state when the loop has finished running
+        if self.done and self.trainer.state.fn != TrainerFn.FITTING:
+            self.dataloader_progress.reset_on_run()
```
Comment (carmocca) [inline pytorch_lightning/loops/dataloader/evaluation_loop.py:87]: I considered putting this in the super implementation, but I don't think the `state.fn` check makes sense there.

**Review by justusschock — APPROVED** [review]

**Review by tchaton — APPROVED** [review]
LGTM !

File: pytorch_lightning/loops/dataloader/evaluation_loop.py (hunk 1 above)
Comment (rohitgr7) [reply pytorch_lightning/loops/dataloader/evaluation_loop.py:90]: quick ques: fault_tolerant doesn't restart/reset the states for `.validate/.test/.predict` call?

Comment (carmocca) [reply pytorch_lightning/loops/dataloader/evaluation_loop.py:90]: It does, see the super implementation.

The bug described here is only for when it is called more than once

**Review by Borda — APPROVED** [review]
